tests/benchmarks/bench_assembly.cpp

```diff
@@ -0,0 +1,389 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "bench.hpp"
+
+#ifdef MFEM_USE_BENCHMARK
+
+/*
+  This benchmark contains the implementation of the CEED's bake-off problems:
+  high-order kernels/benchmarks designed to test and compare the performance
+  of high-order codes.
+
+  See: ceed.exascaleproject.org/bps and github.com/CEED/benchmarks
+
+   * --benchmark_filter=[SetupBP/BP/BK][1-6][PARTIAL/ELEMENT/FULL]/[1-max_order]
+   * --benchmark_context=device=[cpu/cuda/hip]
+*/
+
+// The maximum polynomial order used for benchmarking
+const int max_order = 6;
+// The maximum number of dofs for benchmarking
+const int max_dofs = 1e7;
+
+struct BakeOff
+{
+   const AssemblyLevel assembly;
+   const int N, p, q, dim = 3;
+   Mesh mesh;
+   H1_FECollection fec;
+   FiniteElementSpace fes;
+   const Geometry::Type geom_type;
+   IntegrationRules IntRulesGLL;
+   const IntegrationRule *irGLL;
+   const IntegrationRule *ir;
+   ConstantCoefficient one;
+   const int dofs;
+   GridFunction x,y;
+   BilinearForm a;
+   double mdofs;
+
+   BakeOff(AssemblyLevel assembly, int p, int N, int vdim, bool GLL):
+      assembly(assembly),
+      N(N),
+      p(p),
+      q(2*p + (GLL?-1:3)),
+      mesh(Mesh::MakeCartesian3D(N,N,N,Element::HEXAHEDRON)),
+      fec(p, dim, BasisType::GaussLobatto),
+      fes(&mesh, &fec, vdim),
+      geom_type(fes.GetFE(0)->GetGeomType()),
+      IntRulesGLL(0, Quadrature1D::GaussLobatto),
+      irGLL(&IntRulesGLL.Get(geom_type, q)),
+      ir(&IntRules.Get(geom_type, q)),
+      one(1.0),
+      dofs(fes.GetTrueVSize()),
+      x(&fes),
+      y(&fes),
+      a(&fes),
+      mdofs(0.0) {}
+
+   /// @brief Heuristic to evaluate if the case will run out of memory
+   bool is_runnable() const
+   {
+      const long long int gB = 1073741824/8;
+      const int mem_size = Device::IsEnabled()?16:256;
+      const long long int max_mem = mem_size * gB;
+      const int num_elems = fes.GetNE();
+      long long int mem = num_elems * pow(p+1, dim+1) * 8;
+      if (assembly == AssemblyLevel::ELEMENT)
+      {
+         mem += num_elems * pow(p+1, 2*dim) * 8;
+      }
+      if (assembly == AssemblyLevel::FULL)
+      {
+         mem += 3 * num_elems * pow(p+1, 2*dim) * 8;
+      }
+      // std::cout << "mem = " << mem << " , max_mem = " << max_mem << std::endl;
+      return mem < max_mem;
+   }
+
+   virtual void setup() = 0;
```

Comment:
If this is the benchmark entry point, why do `benchmark` still exists?

Reply:
Because it benchmarks different things, `setup` is to benchmark the "setup", `benchmark` is to benchmark the "apply". I am completely open improving the names and design, I mostly copy pasted what you did here.